storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3109,1760 @@ buf_block_t *fsp_get_latched_xdes_page(
     BUF_GET_POSSIBLY_FREED, mtr, err);
 }
 
+class spaceDefragmenter;
+
+namespace flst
+{
+  /** Validate the file list node for the system tablespace.
+  @param addr file space address
+  @return true if validation successful or false */
+  static bool node_valid(const fil_addr_t *addr) noexcept
+  {
+    return addr->boffset >= FIL_PAGE_DATA &&
+           addr->boffset < (srv_page_size - FIL_PAGE_DATA_END);
+  }
+
+  /** Prepare the steps for removing the file list node
+  @param descr_block descriptor block
+  @param xoffset     descriptor offset within the block
+  @param free_limit maximum free limit in the tablespace
+  @param mtr        mini-transaction
+  @param prev_block previous block in the list
+  @param next_block next block in the list
+  @return error code */
+  dberr_t remove_prepare(const buf_block_t &descr_block,
+                         uint32_t xoffset, uint32_t free_limit,
+                         mtr_t *mtr, buf_block_t **prev_block,
+                         buf_block_t **next_block) noexcept
+  {
+    const xdes_t *descr= descr_block.page.frame + xoffset;
+    fil_addr_t next_addr= flst_get_next_addr(descr);
+    fil_addr_t prev_addr= flst_get_prev_addr(descr);
```

Comment:
In this and other functions, we’d better access the fields in ascending order of address: first the `FLST_PREV` and then the `FLST_NEXT`.